src/monitors/BytecodeProfilingMonitor.v3

```diff
@@ -0,0 +1,450 @@
+// Copyright 2023 Wizard Authors. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+// Implements a profiler that targets wasm modules containing a generalized interpreter
+// that evaluates compiled bytecode.
+//
+// This profiler aims to recover source level information, including hot functions, call graphs,
+// and execution costs in terms of wasm instructions.
+def E = MonitorRegistry.add(
+	"profile_bytecode", "Recovers source level information from bytecode execution",
+	BytecodeProfilingMonitor.new());
+
+def OUT = Trace.OUT;
+
+// the minimum number of cases contained in the "switch" statement of the dispatch
+// function. This is used during static analysis to identify the most likely wasm function
+// as dispatch.
+var MIN_DISPATCH_CASE_COUNT = 100;
+
+// the maximum size in bytes that separates 2 consecutive opcodes.
+var MAX_OPERAND_LEN = 6;
+
+class BytecodeProfilingMonitor extends Monitor {
+	// all functions that could be the dispatch function.
+	// kvp of function index to the (FuncDecl, bytecode_read_pc, profile_state) pair.
+	var potential_dispatch_functions = HashMap<int, (FuncDecl, int, SourceProfileState)>.new(int.!<int>, int.==);
```

Comment:
I think you could replace this with just a list, and have each probe have a reference to its `SourceProfileState`, to avoid a lookup when probes fire.